src/ws/cockpitauth.c

```diff
@@ -1502,11 +1502,20 @@ cockpit_auth_login_async (CockpitAuth *self,
 
   /* If the client sends a TLS certificate to cockpit-tls, treat this as a
    * definitive login type, and don't just silently fall back to other types */
-  const gchar *client_certificate;
   JsonObject *metadata = get_connection_metadata (connection);
-  if (metadata && (client_certificate = json_object_get_string_member (metadata, "client-certificate")))
+  if (metadata && json_object_has_member (metadata, "client-certificate"))
     {
       g_debug ("TLS connection has peer certificate, using tls-cert auth type");
+
+      const gchar *client_certificate = json_object_get_string_member (metadata, "client-certificate");
```

Comment:
I'm not sure it makes sense to do a GError style fail here.  I'd prefer to see a much smaller footprint: maybe just pretend it doesn't exist in that case?

Also: won't json_object_get_string_member throw an assert as well if we get the wrong type?

Reply:
Sure, we could silently ignore type errors, but it's really an unexpected error condition, so I thought it would make sense to fail very verbosely about it, as it's an internal programming error on our side -- possibly when we change the protocol in the future and forget to keep backwards compatibility.

If we get the wrong type, then json_object_get_string_member() [logs a critical](https://gitlab.gnome.org/GNOME/json-glib/-/blob/master/json-glib/json-object.c#L611) (via `g_return_val_if_fail()`), so it'll be similar to what we have in current main. I.e. that's exactly the case when we want to actually fail with "we screwed up the protocol"

Reply:
`cockpit-ws` and `cockpit-tls` are in the same package.  This won't become a compatibility problem in the future.

fwiw, this code is going to get dropped in #16715 anyway, so I don't feel the need to bikeshed too much about it.  If you'd prefer to merge it as-is, I'd be OK with that.  It's definitely an improvement over the current situation.

Reply:
Ack, done. It's simpler that way, and if we are going to change it soon anyway, then indeed no need to be overly picky about it.